### DISCUSSION TITLE
Best BLE proxy for Xiaomi Mijia?

### DISCUSSION THREAD

**Discussion by pzich** [pvvx/ATC_MiThermometer#328] [Q&A] [answered]
I just picked up some Xiaomi Mijia units, and due to the location of my Home Assistant setup, I'm going to need some sort of Bluetooth proxy for my devices to talk to. It seems like [most of the articles](https://esphome.github.io/bluetooth-proxies/) about this suggest using some form of ESP32, but I see your [README](https://github.com/pvvx/ATC_MiThermometer/blob/master/README.md) says not to use ESP32xxx devices:

> Do not use ESP32xxx - it cannot work with BLE at normal speed and drains the batteries of all BLE devices when connected!
>
> ESPHome does not work with Bluetooth 5.0 and misses a lot of advertising packets from BLE devices.

I was wondering if anyone has recommendations for a device that can proxy BLE 4.2 and 5.0+ signals over WiFi? Thank you!

**Comment by AleXSR700** [comment]
Interesting. Never saw that comment on ESP32 devices. Maybe someone knows details on this?
I use a ESP32 running tasmota. No idea if this causes additional battery consumption or not.

**Reply by AleXSR700** [reply]
I wonder if this only applies to devices with default broadcast. I don't quite see how a broadcast every 10 seconds is an issue for the ESP32.

**Reply by pvvx** [reply]
* One antenna for BT and WiFi. Most of the time the antenna is used for Wi-Fi.
* FIFO is not used when receiving BLE advertising.
* Message processing speed is limited by MQTT transaction over WiFi. BLE is not serviced during this period.
* Overloaded software. Slow processing speed of BLE stack on connection. Plus there are a lot of gaps due to Wi-Fi switching. Requires repeated duplication of message transmission for BLE devices. Battery consumption for BLE devices increases tenfold.


With the current bloated codes in the SDK and the rest of the software for simultaneous operation of MQTT-WiFi-BLE, which has long been out of the way of the Flash cache of the ESP32, with 2 cores the ESP32 has performance no higher than Cortex M0 at a frequency of 10..16 MHz. This is due to the speed of reading code from SPI-Flash, which does not fit into the cache. At the same time, the ESP32 heats up thoroughly and consumes more.

**Reply by pvvx** [reply, marked as answer]
> I use a ESP32 running tasmota. 

https://tasmota.github.io/docs/Bluetooth/#bluetooth-low-energy-sensors
_The other advantage is, that scanning for BLE advertisements can happen nearly parallel (very quickly one after the other), while a direct connection must be established for at least a few seconds and **will then block both involved devices** for that time._

With normal BT adapters, no blocking occurs. The BT adapter (client) can simultaneously receive advertising packets and support connection with multiple devices. Also, depending on the software implementation, a BLE device has the ability to simultaneously transmit advertisements and maintain connections with multiple clients.

_ (very quickly one after the other)_ - the required speed of full processing of BLE advertisements from devices is up to 3 ms, which ESP cannot provide. The ESP receives the message and transmits it to MQTT via WiFi. There are delays in WiFi that last more than 1 second. During this time, hundreds of advertising messages from BLE devices will be missed.


WiFi works with a strict step - the default WiFi AP sends a beacon every 102.4 ms. WiFi AP distributes communication time for clients between these beacons.
To reduce collisions in the BLE radio air, the device transmits advertising messages at intervals using a low random delay. But at some time interval, the step to go on air for WiFi and BLE may coincide.
Regular (modern) WIFIs have Bluetooth device detection to reduce collisions. But not ESP.
This is observed as long gaps in data in graphs from users using ESP.

For normal operation, BLE and WiFi antennas must be separated by a distance of half a meter. To work with WiFi you need to use one chip, and to work with BLE - another. Otherwise, there will be a lot of gaps, because the receiver needs to have time to adjust the automatic reception level depending on the noise on the air - to have time to adapt to each BLE advertisement.

https://tasmota.github.io/docs/Bluetooth/#bluetooth-low-energy-sensors
_[pvvx Telink Flasher](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html) also allows to flash custom firmware on [supported sensors](https://github.com/pvvx/ATC_MiThermometer/blob/master/README.md#firmware). This will work out of the box with Tasmota. This firmware does send data more frequently and is a bit **more power hungry than the stock firmware**._

As a result, it turns out that with tasmota, the consumption of the BLE device increases.
With standard BT adapters operating in passive scanning mode, the situation is the opposite - the firmware from MiHome consumes more than alternative firmware.
https://pvvx.github.io/ATC_MiThermometer/OriginalPower.html
https://pvvx.github.io/ATC_MiThermometer/CustPower.html

**Reply by bam80** [reply]
> Requires repeated duplication of message transmission for BLE devices. Battery consumption for BLE devices increases tenfold.

Just to clarify - is this only applies to the bi-directional connection mode?

Tasmota recommends "passive" mode instead which apparently should not suffer: 

> Basically all of them use the so-called "MiBeacons" which are BLE advertisement packets with a certain data structure, which are broadcasted by the devices automatically while the device is not in an active Bluetooth connection. The frequency of these messages is set by the vendor and ranges from one per 3 seconds to one per hour (e.g., for the battery status of the LYWSD03MMC). Motion sensors and BLE remote controls start to send when an event is triggered. These packets already contain the sensor data and can be passively received by other devices and will be published regardless if a user decides to read out the sensors via connections or not. **Thus the battery life of a BLE sensor is not influenced by reading these advertisements and the big advantage is the power efficiency as no active bi-directional connection has to be established.** The other advantage is, that scanning for BLE advertisements can happen nearly parallel (very quickly one after the other), while a direct connection must be established for at least a few seconds and will then block both involved devices for that time.

> This is therefore the preferred option, if supported by the sensor.

https://tasmota.github.io/docs/Bluetooth_ESP32/#mi32-sensors

**Reply by pvvx** [reply]
> Just to clarify - is this only applies to the bi-directional connection mode?

Yes.

[And multiple losses when receive advertising events BLE on ESP](https://github.com/pvvx/ATC_MiThermometer/issues/430) - for other cases.

**Reply by pzich** [reply]
So it sounds like ESP32 is _not_ a good choice, but what _is_? Could I use a raspberry pi, perhaps connected via ethernet and with wifi turned off, so the radio can be used exclusively for bluetooth?

Or what about using its wifi, but with an external bluetooth usb dongle, that could also be placed more than half a meter from the pi itself?

**Comment by pvvx** [comment]
ESP32 cannot successfully handle BLE in Arduino (and over SDK) at 160MHz CLK. Even at 240 MHz, packet acknowledgments are delayed, which greatly affects the batteries of the sensors.

## An example of reading 50 measurements from history in a thermometer.

50 measurements are transferred to the ESP32 in 12 seconds
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/0cf5453a-3194-47df-9309-54b799a3a4d7)
(power СR2032 mA/ms)

ESP32 does not have time to process incoming packets and the thermometer has to constantly duplicate transmissions before receiving confirmations from ESP32. Wi-Fi on the ESP32 is disabled so as not to further slow down BLE.

TelinkMiFlasher.html read 50 measurements (0.2 sec):
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/2126a253-8ea5-4d16-aa79-a858c4dc9913)
(power СR2032 mA/ms)


## Reading the full history from the thermometer’s memory (20 thousand measurements) on any BT adapter (except ESPxx).

ATC_v39, pin-code enable.
Chrome log:
Read time: 18.542 sec

Real read cycle: 17320 ms (by measurement on the graph)

![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/4f9ba550-0312-4adc-b9c6-b690de7cb414)

Average per package: 17320/19632 = 0.8822 ms -> 1133.5 package per second

**Reply by AleXSR700** [reply]
Is this specific to Arduino or would this also apply to tasmota (or some people also use ESPHome)?

And is this only an issue at default transmission times or also when sending at the maximum interval of 10 s?

You commented earlier a graph where the problem only showed when the history was being read. If I am not mistaken (and if I did not misibterpret your graph) then this is never really the case in normal BLE-2-Wifi hub operation.

So if only e.g. 10 BLE sensors are sending data at 10 s intervals, processing time per signal should be longer and if there is a repetition due to wait times for acknowledgment, it should be minimal. Should it not?

P.S.: Is the max interval of 10 s still valid or where you able to find a solution? :)
I could not find whether it was a true hardware limitation or if it was a difficult to solve software issue.

**Reply by pvvx** [reply]
With default settings: An interval of 10 seconds is the sensor polling interval. BLE advertising is transmitted with a period of 2.5 seconds.
As a result, the data message from the sensor is duplicated 4 times and through 3 more communication channels. In BTHome in Home Assistant this will be dots on the graph every 10 seconds.  Each measurement is signed with its own number. The program in BTHome throws out duplicates.
At the same time, there are transmissions from other BLE/MESH devices on the radio.
Some of the messages overlap - a collision occurs. Therefore, duplication of BLE advertising is used.
They arrive at the BT adapter receiver with a frequency of approximately 3 ms. It must process all messages, and select messages that are passed on to MQTT.

BLE advertising of Bluetooth versions prior to 5.4 does not use acknowledgments.
Only a request to send additional information or a connection request is used.
If the “active scanning mode” is enabled, then the BLE device each time transmits an additional packet with the device name and other information. This increases consumption by 25%. 
The request is received after the transmission of the advertising packet, in a window of 500 μs.

On the official website https://www.bluetooth.com/ you can find all the documentation on Bluetooth standards. Bluetooth standards are open to everyone.

> You commented earlier a graph where the problem only showed when the history was being read.

This happens with ESP during any connection with any BLE device.

**Reply by AleXSR700** [reply]
I think you changed default advertising to 1 s in 4.6 firmware. At least mine all reset/changed to 1 s when I performed the update this morning.

Tried setting to 10 s polling but now connecting is hit & (more often) miss.

**Reply by pvvx** [reply]
No. The “Connection delay” was changed to 1 second, several firmware options ago.
This does not apply to the advertising interval.

On the official website https://www.bluetooth.com/ you can find all the documentation on Bluetooth standards. Bluetooth standards are open to everyone.

**Reply by AleXSR700** [reply]
Ah, sorry, my mistake.
Trying to change some settings back.
Do you know why, with Advertising interval set to 10 s, it is extremly difficult to connect via the flasher but the ESP32 continues to get data every 10 s?

**Reply by pvvx** [reply]
Due to errors and limitations in the software of some BT adapters. Where Bluetooth SIG standards do not apply.

The standard specifies a BLE advertising interval of up to 10 seconds.
Windows and Linux require two advertising events to be received to make connection requests. On Android, receiving one advertising event is enough.
The reason is the laziness of software developers for Windows and Linux operating systems.

**Reply by AleXSR700** [reply]
I am using Android.
Hmmm. So in theory setting advertising interval to 10000 ms should work with flasher in Chrome for Android?

**Reply by pvvx** [reply]
Yes. In the nRFConnect program for Android, you can create BLE advertisements in 10-second increments and check.
Also there you can set BLE advertising in the "Coded PHY S8" (LE Long Range) format and check other systems for work (for connection and reception of advertising).

In addition, advertising interval to 10000 ms is not a limitation in the new Bluetooth standards. But we are limited by Bluez and Bleak.

**Reply by AleXSR700** [reply]
Hmm. Could not get it to work on Windows or Android machines with 10 s. Also not with nRFConnect.

**Reply by pvvx** [reply]
https://github.com/pvvx/ATC_MiThermometer/issues/172#issuecomment-1007421408

**Comment by pvvx** [comment]
[Low cost BLE Advertisements Repeater into Ethernet TCP/IP](https://github.com/pvvx/WCHBLE2ETH)

**Reply by pvvx** [reply]
[Low cost BLE Advertisements Repeater into Zigbee](https://github.com/pvvx/TLB2Z)

![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/f8e60511-ddb0-4c15-ae43-9cbc077e1e1c)

BLE advertisements are accepted 99.5% of the time: Zigbee polling occurs 16 ms every 4 seconds. This results in minimal reception loss.

**Comment by AleXSR700** [comment]
> [Low cost BLE Advertisements Repeater into Ethernet TCP/IP](https://github.com/pvvx/WCHBLE2ETH)

Aren't they ~ 20 €? So roughly 5 times the price of an esp32 with wifi?

**Reply by pvvx** [reply]
$0.5 per single chip.
In Russia, in an expensive store, CH32V208W costs $1 if you buy one chip.
Debugger WCH-LinkW with CH32W208W on aliexpress costs $0.5
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/3a7102b8-5f09-4833-b5ab-4f5fbe955464)

There is no ESP device kit with Ethernet from the manufacturer at all.

ESP32 does not work well with BLE. ESP32-C3 is better. And the consumption of ESP+Ethernet is 20 times greater than that of CH32V208W.

**Reply by AleXSR700** [reply]
Admittedly I do not speak russian but I searched for the CH32V208WBU6-EVT-R0 you mentioned and the cheapest one on aliexpress (searching in the european version) is 13 €.
I couls not find any versions witg ethernet for a euro.

I am aware that the esp32 do not offer ethernet, that is why I said wifi :)
Wifi is of course a million times more convenient, especially when the device is supposed to act as a relay/hub.
But I understand your point. RPi is far worse even when it comes to simultaneous use of bluetooth and wifi. Can't even use bluetooth mouse + keyboard while surfing wirelessly. But the esp32 works great as a ble to wifi relay.

But probably, as you mentioned elsewhere in the past, at the cost of battery life of the sensors.

**Reply by pvvx** [reply]
> I am aware that the esp32 do not offer ethernet, that is why I said wifi :)

1. WiFi is not suitable for a Smart Home. I have a hundred devices and I don't want to live in a microwave. 100 BLE devices on average have the same emission as one BLE. Same with Zigbee, but usually Zigbee has a little more power, but not like WiFi. Also, I don’t play with one sensor, but use home automation.
2. Wi-Fi has the lowest reliability. There are many reasons for this. One of them is high consumption and cheap power supplies fail within two years. Cheap Wi-Fi routers do not support hundreds of devices and often freeze. My statistics for 5 years show that on average there is at least 1 connection failure with a WiFi device per day. This can be monitored in Home Assistant logs.
3. Wi-Fi requires powerful routers and each sensor consumes 3 W or more. As a result, my entire smart home network will consume more than 400 W constantly. And this will add air conditioning to cool them.
4. WiFi interferes with Zigbee and BLE. EPS devices with WiFi do not work on 5GHz.
5. Available BLE-WiFi gateway (ESP) implementations have long data transmission intervals. This is not suitable for home automation.
6. BLE-Wifi repeaters on ESP are unstable and have multiple reception dropouts from BLE sensors. That is, they have the lowest level of reliability of all commercially available options. Even if you take BLE routers with WiFi from well-known manufacturers, none of them use ESP chips. The cheap ones use Realtek chips...

No one is stopping you from playing with ESP and a couple of sensors. There are a lot of resources for this, where there are ready-made programs for beginners. ESP software is designed for beginners with the goal of learning and immersing themselves in all the nuances :P

If you need BLE/Zigbee in WiFi, then on aliexpress there is Tuya Smart Gateway BLE Zigbee WiFI [RSH-GW018_DM](https://pvvx.github.io/RSH-GW018_DM/) with modules:
* WBRG1 - RTL872xDM ARM®Cortex™-M4F+M0, 2.5MB RAM, 8MB Flash
* ZS3L - EFR32MG21A020F768IM32-B
2.5MB RAM - This amount of RAM fits a minimal TCP compatible socket with a TCP stack. ESP uses stripped-down ones - used only for toys.
And there are all SDKs for them. You can find out how to flash your programs in such hubs at https://github.com/parasite85/tuya_tygwzw1_hack. But this doesn’t interest me, because I’ve been using such modules for a long time...

EFR32MG21 has good reception sensitivity.

![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/b02465e7-9afd-44cb-a41d-06a7ae829c11)

LILYGO T-Zigbee on ESP32-C3 with TLSR8258 is more expensive, does not have a housing and the characteristics are much worse.

**Reply by AleXSR700** [reply]
I think a lot of people consider ble to be the worst possible choice for Home Automation.

I do not have hundreds of wifi devices but the ~50 I have running in parallel show no signs of failing. Nor does the router seem to have any issues.
Power consumption of a wifi device like a power socket seems to be around 1 W, maybe less, maybe slightly more. Would need a very precise measurement device to check. Not sure if two identical devices but one with ble and one with wifi would have really different consumption (talking mains power, not battery).

The 2.4 Ghz is chosen for the same reason why ble is probably at a disadvantage when it comes to in-wall placement: range.
5 Ghz has poor range through thick walls. And sensors need basically no bandwidth, so 2.4 Ghz is sufficient.
I had to place my esp32 quite carefully to be able to teceive the packages from all bluetooth sensors. And the farthest one is less than 10 m away. Wifi does not present this issue for me.

**Reply by pvvx** [reply]
Write the name of a Wi-Fi router that supports “about 50 parallel connected” Wi-Fi devices.

> Power consumption of a wifi device like a power socket seems to be around 1 W, maybe less, maybe slightly more. 

Place one socket in another and see what it shows :)

> I had to place my esp32 quite carefully to be able to teceive the packages from all bluetooth sensors. 

LE Long Range has a greater transmission range than household WiFi. And no repeaters are required. 
The ESP32 has poor sensitivity for BLE and only supports the legacy Bluetooth standard from 2014.

**Comment by ioukos** [comment]
Hi, 

I'm no expert ; I don't understand why/how, could an ESP32 "listenning" to adverstisment from Xiaomi Sensor drain the battery of the Xiaomi sensor ?

Is it because even if it's an advertisment there is a connexion established between the ESP32 and the sensor ?

I'm more used to 433MHz devices that just "send" their data into the air no matter if another device is listenning.

Anyway, thank you Victor for the good work.

I'm discovering your firmware, and I can read data from ESP32 (through serial) using : 
https://github.com/matthias-bs/ATC_MiThermometer/blob/main/src/ATC_MiThermometer.h

what could be the limitations ? If I run Wifi to send data to MQTT let's say, I won't get all the updates ?

Thank you

**Reply by pvvx** [reply]
> I'm no expert ; I don't understand why/how, could an ESP32 "listenning" to adverstisment from Xiaomi Sensor drain the battery of the Xiaomi sensor ?

We are talking about the connection mode. A full-fledged BLE gateway must be able to connect to devices. For example, to read history or set the time. Possibly OTA too...

And when listening to advertising, if Wi-Fi is on at the same time, there are many reception gaps. And the ESP32 is bad with BLE.
For ESP32 with Bluetooth version 4.2, the average reception range is up to 10 meters.
Even [ESP32 lovers](https://github.com/pvvx/ATC_MiThermometer/discussions/328#discussioncomment-8889892) write about this:
> I had to place my esp32 quite carefully to be able to teceive the packages from all bluetooth sensors. And the farthest one is less than 10 m away. 

 For BLE receiver, it is better to use ESP32-C3 or later. My tests showed [500 meters in BLE LR](https://github.com/pvvx/ATC_MiThermometer/discussions/481#discussioncomment-8822724).

For the WiFi-BLE-Zigbee gateway to work properly, the WiFi and BLE antennas must be at a distance of at least half a meter.

**Comment by BelgarionNL** [comment]
so TLDR whats the best device to get for these bluetooth devices? I still dont see a definitive device I should get from ali express.

**Reply by pvvx** [reply]
This depends on the configuration of your Smart Home.
I use two USB-BT adapters with Home Assistant.
One on the CSR8510A10 works with near 40 sensors, and the second on the RTL8761BU with long-range sensors (covers an area of 500 meters).
For reliability, all this is duplicated by BLE devices of our own assembly. Autonomous control is activated for everyone if the system on Home Assistant stops transmitting alerts or commands at a specified interval.
In addition, most executive devices themselves receive and process data from BLE sensors, and Home Assistant is a recording device for them.
All WiFi routers turn off when operating autonomously due to high consumption. All that remains is a router with Wi-Fi at 5 GHz for communication with the external Internet and local wired Internet to cameras, etc.

The simplest option, published a long time ago https://github.com/pvvx/AdScanerTrg

More complex devices, such as a controller for a water storage tank with support for temperature, fill levels, turning pumps on and off with monitoring of all parameters such as overcurrent, jamming, etc. I don't see the point in publishing it. Because it works differently for everyone.
This is also affected by the weather conditions you live in and the options you enabled when creating your personal living area. For example, in my area of residence in winter, no solar panels work and autonomy is built differently...

**Reply by BelgarionNL** [reply]
ok. I already use a asus bt500. which I believe is the RTL8761BU. its just in use for my pc instead of my unraid server. But I can use it for some testing in my server.

Thanks! I didnt know it could reach 500m.

**Reply by pvvx** [reply]
But such distances are achieved when operating in LE Long Range mode. And while working in LE LR, regular "Legacy" BLE is not accepted. When working on two types of PHY, the adapter periodically switches, which will also contribute to reception failures.
This is why you have to use two adapters. One for BT4.2, and the second in LE LR mode with BT5.0+

In LE mode, the LR RTL8761BU accepts signals down to -110 dBm. 
This is not achievable for chips running Wi-Fi or Zigbee, as well as BT4.2 due to the RF protocols' modulation and restoration techniques. The rest relates to the antennas, not the sensitivity of the chips.

**Reply by 1igeza** [reply]
I have EDUP LOVE EP-B3536 [RTL8761BU] and it's working without any problems with Home Assistant. Can I use second one and set it to LE LR mode? How to select between LE LR and BT4.2 mode?

**Reply by pvvx** [reply]
https://github.com/pvvx/ATC_MiThermometer/issues/297

**Comment by billyl** [comment]
I've been following this thread but I am still confused with the answer that is selected. If my goal is just have a couple of these xiaomi sensors connected to a ESP32-C3 on tasmota within a couple meter range . Is that fine? I am still confused as to what is the recommended device without getting to usb adapters. I would like something that is standalone that can perdiocally log temperatures to a database.

There is a new ESP32C6 that supports zigbee: https://wiki.seeedstudio.com/xiao_esp32c6_getting_started/
Would that be the best alternative?

**Comment by pvvx** [comment]
> There is a new ESP32C6 that supports zigbee: https://wiki.seeedstudio.com/xiao_esp32c6_getting_started/
> Would that be the best alternative?

Try it and write.

**Reply by MisiekDP** [reply]
I bought a ESP32-C6 and wanted to try this BLE but I can't add support with ESPHome. Maybe this module is too much new and doesn't has full support?

**Reply by pzich** [reply]
I believe the C6 is supported now https://esphome.io/changelog/2024.12.0.html

Come by [the discord](https://discord.com/invite/KhAMKrd) if you need help using it.
